e2e/testcases/workload_identity_test.go

```diff
@@ -287,3 +293,52 @@ func truncateStringByLength(s string, l int) string {
 	}
 	return s
 }
+
+// migrateFromGSAtoKSA tests the scenario of migrating from impersonating a GSA
+// to leveraging KSA+WI (a.k.a, BYOID/Ubermint).
+func migrateFromGSAtoKSA(nt *nomostest.NT, rs *v1beta1.RootSync, ksaRef types.NamespacedName, fleetWITest bool) {
+	nt.T.Log("Update RootSync auth type from gcpserviceaccount to k8sserviceaccount")
+	oldGeneration := rs.Generation
+	nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"%s": {"auth": "k8sserviceaccount", "gcpServiceAccountEmail": ""}}}`, rs.Spec.SourceType))
+
+	// Validations
+	nt.T.Log("Validate the GSA annotation is removed from the RootSync's service account")
+	require.NoError(nt.T,
+		nt.Watcher.WatchObject(kinds.ServiceAccount(), ksaRef.Name, ksaRef.Namespace, []testpredicates.Predicate{
+			testpredicates.MissingAnnotation(controllers.GCPSAAnnotationKey),
+		}))
+	if fleetWITest {
+		nt.T.Log("Validate the serviceaccount_impersonation_url is absent from the injected FWI credentials")
+		nomostest.Wait(nt.T, "wait for FWI credentials to exist", nt.DefaultWaitTimeout, func() error {
+			return testutils.ReconcilerPodHasFWICredsAnnotation(nt, nomostest.DefaultRootReconcilerName, "", configsync.AuthK8sServiceAccount)
+		})
+	}
+
+	nt.T.Log("Validate RootSync generation and observedGeneration is updated")
+	require.NoError(nt.T,
+		nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(), rs.Name, rs.Namespace,
+			[]testpredicates.Predicate{
+				testpredicates.GenerationEquals(oldGeneration + 1),
+				testpredicates.HasObservedLatestGeneration(nt.Scheme),
+			}))
+}
+
+func validateAllResourcesReconciled(nt *nomostest.NT, sourceType v1beta1.SourceType, sourceChart, syncDir string, rootCommitFn nomostest.Sha1Func) {
```

Comment:
nit: return error from these functions and invoke fatal from caller

Reply:
Some errors are marked as error (failed, but the test can continue execution), such as validating if a managed resource is reconciled.
Some are marked as fatal and exit immediately, such as CS sync failures.

Handling them inside the function makes it possible to distinguish between the two scenarios. If no strong objections, I would rather keep as is.

Reply:
I tend to prefer failing fast on all such validation errors, in my experience it makes debugging faster/simpler. But no strong objection if you prefer to keep it this way

Reply:
Fair enough. `e2e/nomostest/kustomizecomponents/validate.go` used `nt.T.Error` because it validated each reconciled resource separately, and it was easy to collect all reconcile failures in the past, but we don't see those failures anymore. I also changed them for consistency.